Update GHC to version 8.10.4 and Stackage snapshot to LTS-18.0

aherrmann opened the pull request:
Closes #1464

Updates rules_haskell and all examples and tutorial to GHC version 8.10.4 and Stackage snapshot LTS-18.0.

- Removes `optparse-helper` from `examples/WORKSPACE`. Unfortunately, this package is no longer compatible with LTS-18.0. This was used as a test-case for non-Stackage Hackage dependencies in `stack_snapshot`. We should find a better way to test this feature. https://github.com/tweag/rules_haskell/issues/1561
- Updates some of the vendored example packages for compatibility.
- Updates the dependencies of the persistent worker for compatibility.
- Applies the required changes from https://github.com/tweag/rules_haskell/pull/1464
- Adds GHC to `shell.nix`. This is required by `ghcide`, later versions require `ghc` in `PATH` to determine version and libdir at runtime. Going forward we may want to handle this in the ghcide wrapper. https://github.com/tweag/rules_haskell/issues/1562
- Pull in GHC from a more recent nixpkgs revision.
  Recent GHC require otool and install_name_tool in PATH on MacOS.
  These tools are only included in GHC's PATH in nixpkgs since NixOS/nixpkgs@118b28a.
  This can be undone once we update the main nixpkgs revision. However, this also implies a Bazel update, which I did not want to include in this PR but tackle separately. Therefore, I'm importing a separate nixpkgs revision for GHC for now.
- Override `data-default-instances-containers` and `data-defailt-instances-old-locale` to work around `MAX_PATH` errors on Windows.
    ```
    ghc.exe: loadObj: C:\Users\runneradmin\_bazel_runneradmin\minshlu6\execroot\rules_haskell\bazel-out\x64_windows-fastbuild\bin\external\ghcide\data-default-instances-containers-0.0.1\_install\data-default-instances-containers-0.0.1_iface\HSdata-default-instances-containers-0.0.1.o: file doesn't exist
    ```
    Recent versions of GHC fix many `MAX_PATH` issues on Windows. However, GHC's loader still exposes such an issue. A fix has been merged in GHC HEAD, but is not included in GHC 8.10.4 or 9.0.1. See https://gitlab.haskell.org/ghc/ghc/-/issues/19541
- Set the Cabal option `--disable-library-for-ghci` on Windows to avoid ghcide linking errors of the form
    ```
    unknown symbol `heap_view_closurePtrs'
    ```
    See https://github.com/haskell/ghcide/pull/954
- Set `--remote_default_exec_properties=cache-silo-key=windows-v1` to configure a separate cache silo for Windows.
    One of my attempts to work around the `MAX_PATH` issue above was to configure a shorter Bazel output base. Unfortunately, this poisoned the cache and lead to the following type of build errors:
    ```
    ERROR: D:/a/rules_haskell/rules_haskell/tests/runfiles/BUILD.bazel:20:10: Compiling tests/runfiles/main.cc failed: undeclared inclusion(s) in rule '//tests/runfiles:cc-runner':
   this rule is missing dependency declarations for the following files included by 'tests/runfiles/main.cc':
      '.../execroot/rules_haskell/external/rules_haskell_ghc_windows_amd64/mingw/include/c++/9.2.0/cstdlib'
      '.../execroot/rules_haskell/external/rules_haskell_ghc_windows_amd64/mingw/include/c++/9.2.0/x86_64-w64-mingw32/bits/c++config.h'
    ```
    This only seems to affects builds that include the GHC update. Probably because the cc toolchain on Windows is provided by GHC's bindist. Therefore, I've not split this change out into a separate PR.

I recommend to review this PR commit by commit.

ylecornec reviewed (approved):
It looks good to me, I only have two comments:

- The one in `haskell/ghc_bindist.bzl`
- The documentation for the arm cross compiler in `docs/haskell-use-cases.rst` still mentions version `8.10.2` ([here](https://github.com/tweag/rules_haskell/blob/master/docs/haskell-use-cases.rst?plain=1#L1133) and [here](https://github.com/tweag/rules_haskell/blob/master/docs/haskell-use-cases.rst?plain=1#L1105)).

aherrmann commented on haskell/ghc_bindist.bzl:
```diff
@@ -343,9 +343,16 @@ def _ghc_bindist_impl(ctx):
-        execute_or_fail_loudly(ctx, ["rm", "mingw/lib/gcc/x86_64-w64-mingw32/7.2.0/libstdc++.dll.a"], working_directory = unpack_dir)
         execute_or_fail_loudly(ctx, ["rm", "mingw/x86_64-w64-mingw32/lib/libpthread.dll.a"], working_directory = unpack_dir)
         execute_or_fail_loudly(ctx, ["rm", "mingw/x86_64-w64-mingw32/lib/libwinpthread.dll.a"], working_directory = unpack_dir)
 
```
Good point, yes. I think in those cases the path easier to predict because it didn't contain that mingw version. But, it makes sense to be consistent.

aherrmann commented:
>     * The documentation for the arm cross compiler in `docs/haskell-use-cases.rst` still mentions version `8.10.2` ([here](https://github.com/tweag/rules_haskell/blob/master/docs/haskell-use-cases.rst?plain=1#L1133) and [here](https://github.com/tweag/rules_haskell/blob/master/docs/haskell-use-cases.rst?plain=1#L1105)).

Good catch!

Fixed and rebased on master.

aherrmann commented:
>  GitHub App like Mergify are not allowed to merge pull request where .github/workflows is changed.
>
> This pull request must be merged manually.
